Code/Framework/AzFramework/AzFramework/API/ApplicationAPI.h

```diff
@@ -193,15 +193,29 @@ namespace AzFramework
         : public AZ::EBusTraits
     {
     public:
-        // Bus Configuration
+        static const AZ::EBusHandlerPolicy HandlerPolicy = AZ::EBusHandlerPolicy::Single;
+        static const AZ::EBusAddressPolicy AddressPolicy = AZ::EBusAddressPolicy::Single;
+
+        //! Returns the name of the currently loaded level.
+        //! Note: for spawnable level system, this is the cache folder path to the level asset. Example: levels/mylevel/mylevel.spawnable
+        //! @return Level name or empty string if no level loaded.
+        virtual AZStd::string GetCurrentLevelName() = 0;
+    };
+    using LevelSystemLifecycleRequestBus = AZ::EBus<LevelSystemLifecycleRequests>;
+
+    class LevelLoadBlockerRequests
```

Comment:
This is just such a weird pattern.. arbitrary event handlers that return back to prevent an operation from taking place.. I'm still struggling with it to be honest. It feels like just having a state on the level load system might be a lot less weird?, ie
```
multiplayersystemcomponent::multiplayersystemcomponent()
{
    // LevelLoadSystem is enabled by default since it assumes singleplayer, in multiplayer we need to guard level load operations for correct behaviour
    AZ::Interface<AZ::ILevelLoadSystem>::Get()->SetState(AZ::ILevelLoadSystem::Disabled);
}
...
multiplayersystemcomponent::SetState(state)
{
    if (hosting)
    {
        AZ::Interface<AZ::ILevelLoadSystem>::Get()->SetState(AZ::ILevelLoadSystem::Enabled);
    }
    else
    {
        // client only, or multiplayer disconnect fallthrough
        AZ::Interface<AZ::ILevelLoadSystem>::Get()->SetState(AZ::ILevelLoadSystem::Disabled);
    }
}
```
I still have misgivings about this as well though because it's just so hard to control global state when people go nuts on a codebase for a number of years.

Reply:
Good point... also state setting might need push/pops. One system wants to block while another becomes ready, and so prematurely re-enables the state. Alternatively, we could just allow level loads. There are warnings, but users during workflow session would often load new levels. Thoughts?

Reply:
> Good point... also state setting might need push/pops. One system wants to block while another becomes ready, and so prematurely re-enables the state. Alternatively, we could just allow level loads. There are warnings, but users during workflow session would often load new levels.

I can make the warnings flashier and show them in the viewport; that might be an even better learning opportunity than blocking. "Level load has destroyed your client's player. Try using spawnable api to load in new content instead." Something like that... Thoughts?

Reply:
Hmm, I went to make this a state, and quickly found managing the state is tricky.

Right now Multiplayer blocks state if there are players, are there any "player count changed" event? EnumerateConnections or ConnectionCount wont work since clientserver's dont count themselves in this list.  

The logic currently used...
Clients : always block unless I'm loading the level the server accept message told me to load
ClientServer : always block unless I'm waiting for that 1st level load (the level load that needs to happen right after I call host for myself in order to spawn my own player)
DedicatedServers : level loads are fine, until I get any client players, then dont.

So managing state is actually kind of weird and would require "SetState" in many places. For example, when a client leaves, and I'm a dedicated server, I would then check if all clients are gone, and then ask the level system to set level load state to enabled. 

The current bus that allows systems to handle blocking is nice since it puts all the logic in one function. ShouldBlockLevelLoading() answers that one question.

Reply:
I really like the having all state in one place.. Some additional thoughts could be move the multiplayer agent state to AzFramework somewhere, ComponentApplication potentially? Or create an ILevelFlow interface that could be nullptr and do nothing, or could be non-null and have operations like AllowLevelLoad() or even ProgressToNextLevel() that customer projects could implement and register.. Multiple handler EBus return values always set off alarm bells for me.

If none of those work, then yeah, the benefits of having all the logic together in one easy to audit place are really strong. You shouldn't compromise on that.

Reply:
Definitely seems like MultiplayerGem is the one will needing the level to be blocked, another multiplayer implementation could be different and allow loading all the time. 
Oh, AzFramework might work... it depends on AzNetworking so has access to other stuff I need. Like you mentioned, I would have to move MultiplayerAgent state out of IMultiplayer and into INetworkingInterface(??). Also move NetworkSpawnableFileExtension out of Multiplayer and into AzNetworking (to check what kind of level spawnable is being loaded). From there the spawnable level system could check specifically for all the multiplayer state it needs.

What is rather interesting is all these issues needing to block level load goes away if there was just a way to mark entities as "Do Not Destroy" during level loads. Hmmm.

Reply:
That's something they might do.. however, it's not so simple! 😅 consider two different levels with two different player spawners in each, in the first level spawned players are cube characters, in the second level the spawned players should be sphere characters.. switching levels actually implies destroying and recreating players too, we just need a way to transfer autonomy to the new entity without disconnecting when the old entity is destroyed.